### ISSUE TITLE
Support for Kiiroo Powerblow

### ISSUE THREAD

**Issue by sv-j** [buttplugio/buttplug#578]
**Feature Description**

Are you planning to support Kiiroo's Powerblow? If so do you have any ETA for it? O

Thanks

**Comment by blackspherefollower** [comment]
I've got one on backorder. Support might not be trivial, so no ETA

**Comment by darkfader** [comment]
Maybe estimate the amount of underpressure and go from there. Maybe make it a non-linear function. And probably a low-pass filter since it might not like a lot of pumping/releasing very fast.
It would be nice if it could kinda auto-tune the parameters.

**Comment by SR6Sensation** [comment]
+1! This is totally needed. I am glad to help implementing it but I do not have knowledge of sniffing the Bluetooth protocol.

**Comment by blackspherefollower** [comment]
We have the protocol, but the device isn't an easy fit into buttplug's existing commands: it had 2 commands increase vacuum and release. Mapping those to anything useful isn't trivial.

**Comment by SR6Sensation** [comment]
Does this mean that perhaps Buttplug commands API should be extended to include Suction as a separate parameter instead of mapping it to anything else?

Also interested in how XToys.app has implemented PowerBlow support.

Thanks.

**Comment by blackspherefollower** [comment]
Buttplug already has "Constrict" as a type on ScalarCmd for constant settings on a device, but nothing right now that performs one-off events that the powerblow, the tremblr-bt and the inflatable Satisfyer devices would expect for pressure changes (the last one at least has a concept of a current pressure, whilst the first 2 don't actually know if they're just sucking air from the room Vs the chamber formed between the user and the toy).

**Comment by AhegaHOE** [comment]
It would be very nice to have support for it.

**Comment by AhegaHOE** [comment]
> We have the protocol, but the device isn't an easy fit into buttplug's existing commands: it had 2 commands increase vacuum and release. Mapping those to anything useful isn't trivial.

What about the following behavior and options to configure?

Behavior: The suction should be applied at `suction_strength` when the stroke (depending on `suction_mode`) is at least `minimum_duration` lot and not longer than `maximum_duration`.

Options:

- `suction_strength` integer from 0 to 100 controlling the strength of the vacuum
- `minimum_duration` the minimum duration in ms that a stroke should be for the vacuum to activate
- `maximum_duration` the maximum duration in ms for a single vacuum pull
- `suction_mode` when the vacuum should be applied e.g. on an upward or downward stroke

---

I hope that this helps!

**Comment by qdot** [comment]
It absolutely does not help.

**Comment by AhegaHOE** [comment]
Then what would? I'd be happy to help you guys out with it.

**Comment by blackspherefollower** [comment]
Patience: as I stated previously, there's no good fit for mapping the powerblow's commands to the current Buttplug commands. Adding new commands requires a lot of thought, but something suitable is being planned.

By all means, watch this space, but making guesses about how both the system and the device work and extrapolating those assumptions really isn't useful; it's actually actively misleading to anyone else who stumbles across your comments without understanding the context of the problem.

**Comment by AhegaHOE** [comment]
Alright, thank you. I will be waiting and if any help is needed I would be happy to help!

**Comment by TeKett** [comment]
> Patience: as I stated previously, there's no good fit for mapping the powerblow's commands to the current Buttplug commands. Adding new commands requires a lot of thought, but something suitable is being planned.
> 
> By all means, watch this space, but making guesses about how both the system and the device work and extrapolating those assumptions really isn't useful; it's actually actively misleading to anyone else who stumbles across your comments without understanding the context of the problem.

The most important part is that is just works, and then if it's scuffed you can improve it later.

**Comment by qdot** [comment]
Ok, well, comments closed because not a single one of you has been anywhere near helpful and I'm sick of putting up this. You're welcome for the free software.

To anyone in the future seeing this and wondering what's going on (but are now blocked from getting heated in the comments):

There's a few reasons powerblow implementation is challenging:

- The comments here make me delay it out of spite. Who needs FinDom when you have OSSDom. Tempted to edge people by drafting a powerblow support PR and just leaving it open.
- The powerblow is one of two known devices that do pressure regulation like they do. The other is the pressure cap for the OSR-2/SR-6. With this library generalizing how devices are commanded, building a standard command system around 1.5 devices (I count the OSR-2/SR-6 as 0.5 because it's mostly a DIY deal and an optional add-on for it on top of that, so it's rare we see it in the wild) is very difficult to future proof within our protocol. Our new protocol will make that easier, but that's still in beta, hopefully released soonish (it's been in development for the better part of 2-3 years so there's a lot of work to land, but it's sitting in our dev branch as of this past weekend). We get this wrong, and we're stuck with a wonky one-off implementation that is then stuck because if we change it we break userland, and much like the linux kernel, We Do Not Break Userland. However, unlike the linux kernel, we actually do not break userland.
- The powerblow is an add-on device, made to be run alongside a stroking device, namely the Keon. Every other device we support right now, all ~750 of them, list/have all of their features within the same device. This means we either need to implement compound devices to have a "Keon with Powerblow", or else have to specify two devices that our app developers somehow know to chain together in their user interfaces. This requires messaging to developers that we don't have a good channel for right now. Once our dev docs suck less, that may be different, but that comes as part of (and in this specific case, after) the aforementioned protocol release.

In summary, as we've been saying all along, we're aware people want support but it's complicated. We'll get to it when we get to it, but we have our own schedules and timelines, and we do not care about yours. 

If you don't like it, you are free to start your own sex toy control project. Good luck and godspeed.
